src/sst/core/elemLoader.cc

```diff
@@ -135,92 +56,132 @@ splitPath(const std::string& searchPaths)
-    // on how weak symbol resolution works in dlopen
-    // rerun the loaders to make sure everything is still registered
-    for ( auto& libpair : ELI::LoadedLibraries::getLoaders() ) {
-        // loop all the elements in the element lib
-        for ( auto& elempair : libpair.second ) {
-            // loop all the loaders in the element
-            for ( auto* loader : elempair.second ) {
-                loader->load();
+        if ( nullptr == handle ) {
+            if ( verbose ) { printf("SST-DL: Loading failed, error: %s\n", dlerror()); }
+        }
+        else {
+            if ( verbose ) { printf("SST-DL: Load was successful.\n"); }
+
+            found_element = true;
+
+            for ( auto& libpair : ELI::LoadedLibraries::getLoaders() ) {
+                // loop all the elements in the element lib
+                for ( auto& elempair : libpair.second ) {
+                    // loop all the loaders in the element
+                    for ( auto* loader : elempair.second ) {
+                        loader->load();
+                    }
+                }
             }
+
+            // exit the search loop, we have found the library we tried to load
+            break;
         }
     }
+
+    delete[] full_path;
+
+    if ( !found_element ) { err_os << "Error: unable to find \"" << elemlib << "\" element library\n"; }
+
     return;
 }
 
-extern "C" {
-static int
-elemCB(const char* fname, void* vd)
+void
+ElemLoader::getPotentialElements(std::vector<std::string>& potential_elements)
 {
-    std::vector<std::string>* arr  = (std::vector<std::string>*)vd;
-    const char*               base = strrchr(fname, '/') + 1; /* +1 to go past the / */
-    if ( !strncmp("lib", base, 3) ) {                         /* Filter out directories and the like */
-        arr->push_back(base + 3);                             /* skip past "lib" */
-    }
-    return 0;
-}
-}
+    std::vector<std::string> paths = splitPath(searchPaths);
 
-std::vector<std::string>
-ElemLoader::getPotentialElements()
-{
-    std::vector<std::string> res;
-    lt_dlforeachfile(searchPaths.c_str(), elemCB, &res);
-    return res;
+    for ( std::string& next_path : paths ) {
+        DIR* current_dir = opendir(next_path.c_str());
+
+        if ( current_dir ) {
+            struct dirent* dir_file;
+            while ( (dir_file = readdir(current_dir)) != nullptr ) {
+                // ensure we are only processing normal files and nothing weird
+                if ( (dir_file->d_type | DT_REG) || (dir_file->d_type | DT_LNK) ) {
+                    char* current_file = new char[strlen(dir_file->d_name) + 1];
+                    std::strcpy(current_file, dir_file->d_name);
+
+                    // does the path start with "lib" required for SST
+                    if ( !strncmp("lib", current_file, 3) ) {
+                        // find out if we have an extension
+                        char* find_ext = strrchr(current_file, '.');
```

Comment:
What if you don't have an extension?

Reply:
We are explicitly stating `.so` on Linux and `.so` and `.dylib` on macOS as supported.

Reply:
Okay, I just read the comment `find out if we have ...` as extensions might be optional.